homeassistant/components/recorder/history/modern.py

```diff
@@ -559,40 +617,77 @@ def _get_start_time_state_for_entities_stmt(
     metadata_ids: list[int],
     no_attributes: bool,
     include_last_changed: bool,
+    slow_dependant_subquery: bool,
 ) -> Select:
     """Baked query to get states for specific entities."""
-    # This query is the result of significant research in
-    # https://github.com/home-assistant/core/issues/132865
-    # A reverse index scan with a limit 1 is the fastest way to get the
-    # last state change before a specific point in time for all supported
-    # databases. Since all databases support this query as a join
-    # condition we can use it as a subquery to get the last state change
-    # before a specific point in time for all entities.
-    stmt = (
-        _stmt_and_join_attributes_for_start_state(
-            no_attributes, include_last_changed, False
+    if slow_dependant_subquery:
+        # Simple group-by for MySQL, must use less
+        # than 1000 metadata_ids in the IN clause for MySQL
+        # or it will optimize poorly.
```

Comment:
Should we guard against `metadata_ids` having a length > 1000?

Reply:
I'll add a comment that callers are responsible for ensuring there are less than 1000

Reply:
2cb7d00217f